File: packages/py/analytics/src/bolsa_analytics/cognitive/exit_risk_signature.py
"""V1.32 — firma de tamaño en salida SEMI (simétrico a risk_signature)."""

from __future__ import annotations

from typing import Any

QTY_EPS = 1e-9


def _finite(value: object) -> float | None:
    if isinstance(value, bool) or not isinstance(value, int | float):
        return None
    n = float(value)
    if n != n or abs(n) == float("inf"):
        return None
    return n


def _audited(reason: object | None) -> bool:
    return isinstance(reason, str) and bool(reason.strip())


def evaluate_exit_risk_signature(
    *,
    planned_qty: float | None,
    signed_qty: float,
    override_reason: str | None = None,
) -> dict[str, Any]:
    """Qty firmada ≤ planned; exceso exige override auditable."""
    signed = _finite(signed_qty)
    if signed is None or signed <= 0:
        planned = _finite(planned_qty)
        return {
            "mode": "exit",
            "plannedQty": planned,
            "maxQty": planned,
            "overrideRequired": False,
            "allowed": False,
            "excess": None,
            "blockReason": "qty_invalid",
        }

    planned = _finite(planned_qty)
    if planned is None or planned <= 0:
        return {
            "mode": "no_plan",
            "plannedQty": None,
            "maxQty": None,
            "overrideRequired": False,
            "allowed": True,
            "excess": None,
            "blockReason": None,
        }

    excess = signed - planned
    exceeds = excess > QTY_EPS
    override = _audited(override_reason)
    return {
        "mode": "exit",
        "plannedQty": planned,
        "maxQty": planned,
        "overrideRequired": exceeds,
        "allowed": (not exceeds) or override,
        "excess": excess if exceeds else None,
        "blockReason": "qty_exceeds_plan" if exceeds and not override else None,
    }

File: packages/shared/src/cognitive/exit-risk-signature.ts
/**
 * V1.32 — firma de tamaño en salida SEMI (simétrico a risk_signature de apertura).
 * Qty firmada ≤ plannedQty del ExitPlan / enqueue; exceso exige override.
 * ≠ check_opening · ≠ TradePlan · ≠ ExitPermission.
 */

export const EXIT_QTY_EPS = 1e-9;

export type ExitRiskSignatureV1 = {
  mode: "exit" | "no_plan";
  plannedQty: number | null;
  maxQty: number | null;
  overrideRequired: boolean;
  allowed: boolean;
  excess: number | null;
  blockReason: "qty_exceeds_plan" | "qty_invalid" | null;
};

function finite(value: unknown): number | null {
  if (typeof value !== "number" || !Number.isFinite(value)) return null;
  return value;
}

function auditedReason(reason: unknown): boolean {
  return typeof reason === "string" && reason.trim().length > 0;
}

export function evaluateExitRiskSignature(input: {
  plannedQty: number | null | undefined;
  signedQty: number;
  overrideReason?: string | null;
}): ExitRiskSignatureV1 {
  const signed = finite(input.signedQty);
  if (signed == null || signed <= 0) {
    return {
      mode: "exit",
      plannedQty: finite(input.plannedQty),
      maxQty: finite(input.plannedQty),
      overrideRequired: false,
      allowed: false,
      excess: null,
      blockReason: "qty_invalid",
    };
  }

  const planned = finite(input.plannedQty);
  if (planned == null || planned <= 0) {
    // Sin plan de qty (legado): no bloquea por firma de tamaño.
    return {
      mode: "no_plan",
      plannedQty: null,
      maxQty: null,
      overrideRequired: false,
      allowed: true,
      excess: null,
      blockReason: null,
    };
  }

  const excess = signed - planned;
  const exceeds = excess > EXIT_QTY_EPS;
  const override = auditedReason(input.overrideReason);

  return {
    mode: "exit",
    plannedQty: planned,
    maxQty: planned,
    overrideRequired: exceeds,
    allowed: !exceeds || override,
    excess: exceeds ? excess : null,
    blockReason: exceeds && !override ? "qty_exceeds_plan" : null,
  };
}
